feat(import): add import flow component

vikram-raj opened the pull request:
Story - https://jira.coreos.com/browse/ODC-301

Bumbilo committed:
feat(import_flow_component): Create simple UI for import flow
- Create React component for import flow

Bumbilo committed:
feat(import_flow_component): Create simple UI for import flow
- Create React component for import flow

Bumbilo committed:
feat(import_flow_component): Create simple UI for import flow
- Create React component for import flow

vikram-raj committed:
[WIP]feat(import): add import flow component

vikram-raj committed:
[WIP]feat(import): add import flow component

vikram-raj committed:
[WIP]feat(import): add import flow component

vikram-raj committed:
fix(import): remove mock catalog component

vikram-raj committed:
fix(import): remove mock catalog component

vikram-raj committed:
fix(import): remove mock catalog component

vikram-raj committed:
fix(import): add form component from patternfly

vikram-raj committed:
fix(import): add form component from patternfly

vikram-raj committed:
fix(import): add form component from patternfly

vikram-raj committed:
fix(import): connect import-flow component with redux
 - fetch project name and show in application name dropdown
 - fix close modaloverlay issue.

vikram-raj committed:
fix(import): connect import-flow component with redux
 - fetch project name and show in application name dropdown
 - fix close modaloverlay issue.

vikram-raj committed:
fix(import): connect import-flow component with redux
 - fetch project name and show in application name dropdown
 - fix close modaloverlay issue.

vikram-raj committed:
fix(import): remove application mock object

christianvogt commented on frontend/public/extend/devconsole/pages/Import.tsx:
```diff
@@ -1,7 +1,15 @@
+  <div className='import-container'>
+    <Firehose resources={[{kind: NamespaceModel.kind, prop: 'namespace', isList: true}]}>
+      <ImportFlow />
+    </Firehose>
```
You should have another component that encapsulates this `Firehose` and `ImportFlow` together.

christianvogt commented on frontend/public/components/namespace.jsx:
```diff
@@ -256,6 +264,11 @@ class NamespaceBarDropdowns_ extends React.Component {
 
+  importOverlay() {
+    const isOpen = this.state.isOpen;
+    this.setState({isOpen: !isOpen});
```
When setting the state based on the existing state value, always use the function parameter form of the `setState` function:
```
this.setState(({isOpen}) => ({isOpen: !isOpen}));
```

christianvogt commented on frontend/public/components/namespace.jsx:
```diff
@@ -247,6 +248,13 @@ const namespaceBarDropdownStateToProps = state => {
+  constructor(props) {
+    super(props);
+    this.state = {
+      isOpen: false,
```
`isOpen` is too generic. What is open? Make it more specific.

christianvogt commented on frontend/public/components/namespace.jsx:
```diff
@@ -298,6 +315,8 @@ class NamespaceBarDropdowns_ extends React.Component {
 
     return <div className="co-namespace-bar__dropdowns">
+      <PreviewImport
+        isOpen={this.state.isOpen} closeModal={this.importOverlay} />
```
I'm really not sure if this is the best approach to showing the modal. But I know that the new designs are moving away from this dropdown so I'm not too concerned at this time.

christianvogt commented on frontend/public/extend/devconsole/pages/import.scss:
```diff
@@ -0,0 +1,4 @@
+.import-container {
+  max-width: 700px;
+  margin: 30px auto;
+}
```
Follow css naming conventions. Use prefix `odc-`. Name the rule to match the component name. Any element inside the component should use the notation `__element-name`.

Also use patternfly variables for sizing. IIRC patternfly does spacing in increments of `8px`.
Probably do not need the margin if you're using `PageSection` from patternfly.

Why are you constraining the width to `700px`?

vikram-raj replied:
I tried ```PageSection``` from Patternfly but it doesn't add any space from the top.

I fix max-width to 700px so, that I can align the form to the center of the page. I will change it once we have the final UX design.

christianvogt replied:
Damn console overrides the patternfly styles :(

christianvogt commented on frontend/public/extend/devconsole/components/preview-import/preview-import.scss:
```diff
@@ -0,0 +1,5 @@
+  &__body {
+    padding: 20px 50px 30px 20px;
+  }
+}
```
Follow css naming conventions. Use prefix odc-. Name the rule to match the component name. Any element inside the component should use the notation __element-name.

Also use patternfly variables for sizing. IIRC patternfly does spacing in increments of 8px.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+}
+const mapStateToProps = (state) => {
+  return state;
+}
```
Missing `DispatchProps` for correct typing.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+import { NamespaceModel, ProjectModel } from './../../../../../public/models';
+import { connect } from 'react-redux';
+
+class ImportFlow extends React.Component<any, any> {
```
Do not use `any` type. Use proper typings.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+  Button
+} from '@patternfly/react-core';
+import { NamespaceModel, ProjectModel } from './../../../../../public/models';
+import { connect } from 'react-redux';
```
Organize imports absolute first, followed by relative, followed by scss.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow-catalog.tsx:
```diff
@@ -0,0 +1,30 @@
+import * as React from 'react';
+import { CatalogTile } from 'patternfly-react-extensions';
+
+export const ImportFlowCatalog: React.SFC<any> = props => {
```
Do not use `any` typing. Use proper types.
We prefer default exports for react component files to ensure we stick to a single exported component per file.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow-catalog.tsx:
```diff
@@ -0,0 +1,30 @@
+export const ImportFlowCatalog: React.SFC<any> = props => {
+  return (
+    <div className="catalog-tile-view-pf catalog-tile-view-pf-no-categories import__catalog">
+      <div className="import__filter">
```
Where are these classes defined?
This entire render function seems odd. Shouldn't there be reusable components? Why are we using classes for styling here?

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+
+  render() {
+    const { namespace } = this.props;
+    return (
```
Deconstruct upfront all props and state that will be used in the render function.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+          <p>
+            Some help text about the section lorem ipsum
+          </p>
+        </div>
```
Remove the wrapping `<div>`

christianvogt commented on frontend/public/extend/devconsole/components/preview-import/preview-import.tsx:
```diff
@@ -0,0 +1,39 @@
+import ImportFlow from './../import/import-flow';
+import './preview-import.scss';
+import { Firehose } from './../../../../../public/components/utils';
+import { NamespaceModel } from './../../../../../public/models';
```
Organize imports absolute first, followed by relative, followed by scss.
Also fix your relative paths to the shortest possible path.

christianvogt commented on frontend/public/extend/devconsole/components/preview-import/preview-import.tsx:
```diff
@@ -0,0 +1,39 @@
+export class PreviewImport extends React.Component<any, any> {
+  constructor(props) {
+    super(props);
+  }
```
This should be a `FunctionalComponent` since there's no state. Also missing typings. Do not use `any`.

christianvogt commented on frontend/public/extend/devconsole/components/preview-import/preview-import.tsx:
```diff
@@ -0,0 +1,39 @@
+            </Firehose>
+          </div>
+        </Modal.Body>
+      </ModelessOverlay>
```
As much as I dislike their approach right now, OpenShift console has a way of doing modals using `createModalLauncher` from `factory/modal`.
Have a look.

christianvogt commented on frontend/public/extend/devconsole/components/preview-import/preview-import.tsx:
```diff
@@ -0,0 +1,39 @@
+          <div className='preview-import__body'>
+            <Firehose resources={[{kind: NamespaceModel.kind, prop: 'namespace', isList: true}]}>
+              <ImportFlow />
+            </Firehose>
```
You should have another component that encapsulates this `Firehose` and `ImportFlow` together. Or just put the `Firehose` in the `ImportFlow`.

vikram-raj replied:
I separated the import-flow component into two. And put the ```Firehose``` at one place in ```importFlow``` component.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+      name: '',
+      builderImage: ''
+    };
+  }
```
If you separated out this component into two, one for managing state and another for rendering, you'll have an easier time testing the two components. It would also align better with the redux approach and would later allow you to more easily switch state management from local state to redux or something else.
`ImportFlow` and `ImportForm`

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,147 @@
+  FormSelect,
+  ActionGroup,
+  Button
+} from '@patternfly/react-core';
```
I think for consistency we need to stick to the PF3 react components. Those would align with the mock ups creates by UX.

vikram-raj replied:
```FormSelect```, ```FormSelectOption```, ```FormAction```, ```TextInput``` these component are not present in PF3 react component.

christianvogt replied:
[PF3 vertical form example](http://patternfly-react.surge.sh/patternfly-3/index.html?knob-Title=Kebab%20title&knob-Show%20Modal=true&knob-Close%20Popover=true&knob-Show%20Help=true&knob-Field%20Level%20Help%20Content=Please%20specify%20Country%20code%20%3Cbr%3E%20%3Ca%20target%3D%27_blank%27%20href%3D%27https%3A%2F%2Fcountrycode.org%2F%27%3EClick%20here%20for%20a%20list%20of%20Country%20codes%3C%2Fa%3E&knob-Leeway=2&knob-Max%20Shown=5&selectedKind=patternfly-react%2FForms%20and%20Controls%2FForms%2FForm%20Examples&selectedStory=Vertical%20Form&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)

For the select, you can use `components/utils/dropdown` from console, or PF3 typeahead select, or PF3 Dropdown. Probably the console dropdown is your best choice to align with the rest of the console.

vikram-raj replied:
I replace the form with PF3 for. and Also added typeahead dropdown to application name dropdown.

christianvogt reviewed (changes requested):
Run prettier on all your files in the `devconsole` folder.
Run the linter.
Follow file naming conventions in `devconsole` folder when naming component classes. Uppercase first letter and camel casing to match the name of the component default export.

vikram-raj committed:
fix(import): add builder image field

vikram-raj committed:
fix(import): PR review changes
 - Follow BEM notation for class name
 - Rename preview-import -> OverlayImportFlow
 - Change OVerlayImportImport component from class to functional component
 - Re-arrange the dependecies import

vikram-raj committed:
fix(import): PR review changes
 - Follow BEM notation for class name
 - Rename preview-import -> OverlayImportFlow
 - Change OVerlayImportImport component from class to functional component
 - Re-arrange the dependecies import

vikram-raj committed:
fix(import): PR review changes
 - Follow BEM notation for class name
 - Rename preview-import -> OverlayImportFlow
 - Change OVerlayImportImport component from class to functional component
 - Re-arrange the dependecies import

vikram-raj committed:
fix(import): make separate component for import flow form.

vikram-raj committed:
fix(import): make separate component for import flow form.

vikram-raj committed:
fix(import): make separate component for import flow form.

sahil143 commented on frontend/public/components/namespace.jsx:
```diff
@@ -256,6 +264,10 @@ class NamespaceBarDropdowns_ extends React.Component {
     }
   }
 
+  importOverlay() {
```
shouldn't `property intializer syntax` also used here ? `importOverlay = () => { }`

sahil143 commented on frontend/public/extend/devconsole/pages/Import.tsx:
```diff
@@ -1,7 +1,11 @@
+import { ImportFlow } from './../components/import/import-flow';
+import './import.scss';
 
 const ImportPage: React.SFC = () => (
```
Please change the type `SFC` to `FunctionComponent`?

sahil143 commented on frontend/public/extend/devconsole/pages/Home.tsx:
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react';
+import { Button } from '@patternfly/react-core';
 
 const HomePage: React.SFC = () => {
```
Please change the type SFC to FunctionComponent? Because SFC is deprecated.

vikram-raj committed:
fix(import): use PF3 form in ImportFlowForm component
 - add typeahead dropdown for application name field

vikram-raj committed:
fix(import): use PF3 form in ImportFlowForm component
 - add typeahead dropdown for application name field

vikram-raj committed:
fix(import): use PF3 form in ImportFlowForm component
 - add typeahead dropdown for application name field

vikram-raj committed:
fix(import): remove mock data for builder image

vikram-raj committed:
fix(import): remove mock data for builder image

vikram-raj committed:
fix(import): remove mock data for builder image

vikram-raj committed:
fix(import): use functionaComponent instead of SFC

vikram-raj committed:
fix(import): show import from git option in secondary header add option only in dev dev prespective

vikram-raj committed:
fix(import): show import from git option in secondary header add option only in dev dev prespective

vikram-raj committed:
fix(import): show import from git option in secondary header add option only in dev dev prespective

christianvogt commented:
@vikram-raj the `+ Add` menu is going away entirely. It'll be replaced with the `+ Add` main nav item and a landing page. So please remove your changes to the project bar.

vikram-raj committed:
fix(import): add URL validation and git type detection

vikram-raj committed:
fix(import): add URL validation and git type detection

vikram-raj committed:
fix(import): add URL validation and git type detection

vikram-raj commented:
@christianvogt ```+ Add``` option from main nav will open the entire page. If we remove the ```+ add``` from the project bar then from where import flow slider will open.

christianvogt commented:
`+ Add` will open a landing page with all the same options as those that appear in the dropdown menu today but will be a tile selector.

vikram-raj commented:
So, when we need the import flow that will slide in from the left?

christianvogt commented:
Right now the only import flow that is has discussed is via the topology empty state and from the `+ Add` main navigation item. Both result in the same landing page content and both open import flow in main content area.

There may be future flows in a panel. But the drop-down menu that you've modified is going away.

vikram-raj commented:
Got it. Thanks

vikram-raj committed:
fix(import): remove import from git option from add dropdown

vikram-raj committed:
fix(import): remove import from git option from add dropdown

vikram-raj committed:
fix(import): remove import from git option from add dropdown

vikram-raj committed:
fix(import): remove overlayImport state

vikram-raj committed:
fix(import): remove overlayImport state

vikram-raj committed:
fix(import): remove overlayImport state

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,226 @@
+}
+
+const mapStateToProps = (state) => {
+  return state;
```
Never pass in the entire state otherwise your component re-renders on every state change that may not even be relevant to your component.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,226 @@
+      gitTypeField = null;
+    }
+    return (
+      <div className="odc-import-flow-form">
```
The form itself shouldn't assign padding. Let the parent who consumes the form own the padding.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,226 @@
+      <div className="odc-import-flow-form">
+        <p>
+          Some help text about the section lorem ipsum
+        </p>
```
Same comment about the info text. It probably shouldn't be in the form itself. Try to make components that are very focused and reusable.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,226 @@
+          onChange={this.handleGitTypeChange} />
+      </FormGroup>;
+    } else {
+      gitTypeField = null;
```
This else is unnecessary.

christianvogt commented on frontend/public/extend/devconsole/pages/Import.tsx:
```diff
@@ -1,7 +1,11 @@
+const ImportPage: React.FunctionComponent = () => (
+  <div>
+    <ImportFlow />
+  </div>
```
Update this page to follow the standard page layouts. See other `*-page.tsx` files.

christianvogt commented on frontend/public/extend/devconsole/pages/Home.tsx:
```diff
@@ -9,6 +10,9 @@ const HomePage: React.SFC = () => {
+      </Button>
     </div>
   );
 };
```
Home page is now gone. 
Add a new route for your import flow.

christianvogt commented on frontend/public/extend/devconsole/components/overlay-import-flow/OverlayImportFlow.tsx:
```diff
@@ -0,0 +1,34 @@
+    </ModelessOverlay>
+  );
+};
+
```
You won't need this modal anymore with the full page view.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,226 @@
+  namespace: {
+    data: Array<any>,
+  }
+}
```
Can we get a better description of the prop types instead of using `any`

vikram-raj committed:
fix(import): add tests for ImportFlowFormComponent

vikram-raj committed:
fix(import): add unit test for import flow form submission

vikram-raj committed:
fix(import): add unit test for import flow form submission

vikram-raj committed:
fix(import): remove style from ImportFlowForm component and add to parent compnent
 - remove unnecessary else condition

vikram-raj committed:
fix(import): remove style from ImportFlowForm component and add to parent compnent
 - remove unnecessary else condition

vikram-raj committed:
fix(import): delete overlay import component

vikram-raj committed:
fix(import): delete overlay import component

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,243 @@
+              required
+              value={gitRepoUrl}
+              onChange={this.handleGitRepoUrlChange}
+              onBlur={this.validateGitRepo}
```
Validate the git url against the regex `onChange`.
Validation via the CR will happen on blur.

christianvogt commented on frontend/public/extend/devconsole/components/import/import-flow.tsx:
```diff
@@ -0,0 +1,22 @@
+      </h1>
+      <p>
+        Some help text about the section lorem ipsum
+      </p>
```
I think that should be part of the page consuming the flow.
There shouldn't be any padding around this component. Leave that to the parent.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,243 @@
+            <Button type="button">Cancel</Button>
+          </div>
+        </Form>
+      </div>
```
No need for this outter `div`.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,243 @@
+      this.setState({ gitType: 'gitlab', gitTypeDetected: true });
+    } else {
+      this.setState({ gitType: '', gitTypeDetected: true });
+    }
```
Since `setState` is async. It's best to only have a single call to `setState` to ensure it's clear that no part of the flow is consuming state and expecting a new state value.
Cleaner to assign two variables `gitType` and `gitTypeDetected`. And then call `setState` once.
Or better yet, have the `detectGitType` as a static function that simply returns the git type detected, then set the state with this value in `validateGitRepo`.

vikram-raj committed:
fix(import): addressed review comment
 - change detectGitType function type
 - validate valid URL onChange
 - add namespace bar

vikram-raj committed:
fix(import): addressed review comment
 - change detectGitType function type
 - validate valid URL onChange
 - add namespace bar

vikram-raj committed:
fix(import): move import flow info section to parent component

vikram-raj committed:
fix(import): move import flow info section to parent component

Commit:
feat(import-flow): validate git url

vikram-raj committed:
fix(import): add cancle button functionality

vikram-raj committed:
fix(import): add model for namespace

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,272 @@
+  handleCancel = (event) => {
+    event.preventDefault();
+    this.setState(initialState);
+    history.back();
```
import `history` from `frontend/public/components/utils/router.ts`
Then use `history.goBack()`

vikram-raj committed:
fix(import): use history function from utils/router

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,273 @@
+            onChange={this.handleApplicationNameChange}
+            autocompleteFilter={this.autocompleteFilter}
+            autocompletePlaceholder={'Select application name'}
+            data-test-id="import-application-name" />
```
Can you use `NsDropdown` from `utils`?
Also the label isn't `Application Name`. It should be `Namespace`.

vikram-raj committed:
fix(import): add helmet component to import flow page
 - validate URL in UI after 2sec timeout

debsmita1 committed:
feat(import-flow):Clean up temporary git source

vikram-raj committed:
fix(import): allow to select the git type when git type not detected

debsmita1 committed:
feat(import-flow):Clean up git source on browser kill/reload

vikram-raj committed:
fix(import): disable create button if required field is not filled

vikram-raj committed:
fix(import): add create button functionality

vikram-raj committed:
fix(import): rename application field to namespace and us NsDropdown instead of Dropdown

vikram-raj commented:
@christianvogt I added @debsmita1 changes to this PR as well. Please have a look.

vikram-raj committed:
fix(import): set crd true for GitsourceModel and GitsourceComponentModel

vikram-raj committed:
fix(import): lint error

divyanshiGupta commented on frontend/public/extend/devconsole/routes.tsx:
```diff
@@ -5,7 +5,7 @@ import { AsyncComponent } from '../../components/utils';
 const routes: RouteProps[] = [
   {
-    path: '/dev/add',
+    path: '/dev/k8s/ns/:ns/add',
```
Shouldnt we create a separate nav-item for import and a separate route? Add is supposed to show only the empty state at all times with the `+add` actions.

vikram-raj replied:
Yes, I will add a separate route for import.

divyanshiGupta commented on frontend/public/extend/devconsole/routes.tsx:
```diff
@@ -16,6 +16,18 @@ const routes: RouteProps[] = [
     ),
   },
+  {
+    path: '/dev/k8s/ns/:ns/git-import',
```
@vikram-raj can you change the path to `/dev/git-import/ns/:ns` . Refer this comment https://github.com/talamer/console/pull/38#discussion_r277704208 by @christianvogt .

divyanshiGupta replied:
Also, you will have to redirect `/dev/git-import` to `/dev/git-import/ns/:ns` or `/dev/git-import/all-namespaces`. Refer latest routes.tsx.

sbose78 commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+  };
+
+  builderImages = {
+    '': 'Please choose builder image',
```
Ideally you should be fetching this list from the equivalent of `oc get imagestreams -n openshift`

vikram-raj committed:
fix(import): display error when user try to create componet with same

christianvogt commented on frontend/public/models/index.ts:
```diff
@@ -40,6 +40,32 @@ export const CatalogSourceModel: K8sKind = {
+  namespaced: true,
+  kind: 'Component',
+};
+
```
We should probably put these models in our folder for now.

debsmita1 replied:
Done https://github.com/vikram-raj/console/pull/3

christianvogt commented on frontend/public/extend/devconsole/pages/Import.tsx:
```diff
@@ -1,7 +1,24 @@
+      </h1>
+      <p>
+        Some help text about the section lorem ipsum
+      </p>
```
Remove any temp text for now so when we merge we don't have this in any demos.

Console uses the `PageHeading` component. Follow what the rest of the pages do.

debsmita1 replied:
Should we remove the helper texts, below the input fields, as well ? @christianvogt

debsmita1 replied:
https://github.com/vikram-raj/console/pull/3

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/__tests__/ImportFlowForm.spec.tsx:
```diff
@@ -0,0 +1,90 @@
+    });
+    importFlowForm.find('input[data-test-id="import-git-repo-url"]').simulate('blur');
+    importFlowForm.find('Dropdown[data-test-id="import-application-name"]').simulate('change');
+    importFlowForm.find('input[data-test-id="import-name"]').simulate('change', {
```
You should only need the test id as part of your query. No need for the tag to also be in the selector.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+        const urlRegex = /^(?:http(s)?:\/\/)?[\w.-]+(?:\.[\w.-]+)+[\w\-._~:/?#[\]@!$&'()*+,;=.]+$/;
+        if (!urlRegex.test(this.state.gitRepoUrl)) {
+          this.setState({ gitRepoUrlError: 'Please enter the valid git URL' });
+          this.setState({ gitType: '' });
```
Call `setState` only once.

debsmita1 replied:
Done https://github.com/vikram-raj/console/pull/3

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+    let timeOut;
+    this.setState({ gitRepoUrl: event.target.value });
+    clearTimeout(timeOut);
+    timeOut = setTimeout(() => {
```
Timeout must be saved as an instance variable. It should also be cleared on `componentWillUnmount`
Is this even necessary? It's just doing regex comparison we can do `onChange` fast enough.

vikram-raj replied:
Yes, it is not actually necessary. Because we are not making any API call here to validate the URL, just checking with regex.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+
+  componentDidMount() {
+    const activeNamespace = getActiveNamespace();
+    this.randomString = this._generateRandomString();
```
Move random string initialization to the declaration itself.

debsmita1 replied:
Done https://github.com/vikram-raj/console/pull/3

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+  }
+
+  componentDidMount() {
+    const activeNamespace = getActiveNamespace();
```
We should get the active namespace from redux connect and not from some global function.

debsmita1 replied:
Done https://github.com/vikram-raj/console/pull/3

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+  gitSourceCreated: boolean,
+  resourceVersion: string,
+  lastEnteredGitUrl: string,
+  componentCreated: boolean,
```
A lot of types here could be marked optional such as the errors so you can not use `null` as a value instead of empty string.

christianvogt commented on frontend/public/extend/devconsole/components/import-flow-form/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,410 @@
+  private randomString = '';
+
+  // handles cases where user reloads the form/closes the browser
+  private _onBrowserClose = event => {
```
we don't need to prefix private functions with underscore in typescript.

debsmita1 replied:
Done https://github.com/vikram-raj/console/pull/3

christianvogt reviewed (dismissed):
Also, update your file names for components to use PascalCase.

debsmita1 committed:
Incorporating review comments

sbose78 commented:
@vikram-raj , could we show the namespace context in the Import page?

vikram-raj commented:
@sbose78 Yes, we show the namespace context.

vikram-raj committed:
fix(import): change import URL

vikram-raj committed:
fix(import): update with master-next

debsmita1 committed:
Modified implementation of GitSource Creation

vikram-raj committed:
fix(import): remove setTimeout from regex check

debsmita1 committed:
feat(import-flow):read git url status from GS_CR

debsmita1 committed:
feat(import-flow):read git url status from GS_CR

christianvogt commented on frontend/public/extend/devconsole/pages/Import.scss:
```diff
@@ -0,0 +1,3 @@
+.odc-import-page-wrapper {
+  padding: 0px 20px 20px 30px;
+}
```
Remove this and use `co-m-pane__body` instead.

christianvogt commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,395 @@
+  }
+
+  gitTypes = {
+    '': 'please choose Git type',
```
`please` -> `Please`
Also `Git` case doesn't match `git` used elsewhere.

christianvogt commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,395 @@
+            id="import-git-repo-url"
+            data-test-id="import-git-repo-url"
+            name="gitRepoUrl" />
+          {showGitValidationStatus}
```
This doesn't show up well on mobile with the padding changes suggested:
![image](https://user-images.githubusercontent.com/14068621/56770094-1ad32c80-6781-11e9-8ef8-23f27626dc1e.png)
Should we not have the loader and validation beside the label?

vikram-raj replied:
Moved it to beside the label.

christianvogt commented on frontend/public/extend/devconsole/components/import/ImportFlow.tsx:
```diff
@@ -0,0 +1,14 @@
+import { Firehose } from '../../../../components/utils';
+import { NamespaceModel } from '../../../../models';
+import ImportFlowForm from '../ImportFlowForm/ImportFlowForm';
+import './ImportFlow.scss';
```
Doesn't look like you need `ImportFlow.scss`.

christianvogt commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,395 @@
+      return 'gitlab';
+    }
+    this.setState({ gitTypeError: 'Not able to detect the git type. Please choose git type' });
+    return '';
```
This is weirdly done as the caller of `detectGitType` sets state and here we also set state. Choose one location to set the state.

vikram-raj replied:
Fix this. Remove the ```setState``` from ```detectGitType``` function.

christianvogt commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,395 @@
+    history.goBack();
+  }
+
+  checkUrlValidationStatus() {
```
Change to `checkUrlValidationStatus: () => {` so you can avoid using `.bind`

vikram-raj replied:
I missed this.

christianvogt commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx:
```diff
@@ -0,0 +1,395 @@
+            }`,
+            lastEnteredGitUrl: this.state.gitRepoUrl,
+          });
+          this.poller=setInterval(this.checkUrlValidationStatus.bind(this), 3000);
```
Why is this an interval and not a timeout?
Why not watch with a web socket? Why poll?

vikram-raj replied:
> Why is this an interval and not a timeout?

Because we have to keep checking the GitSource until we got the status of the GitSource.

>Why not watch with a web socket? Why poll?

We tried to use existing ```watchK8sObject``` but we are facing some issue with it.

christianvogt replied:
Looks like on the first interval callback to `checkUrlValidationStatus` and the very first result returned by `k8sGet` the interval is cancelled. So it doesn't look like you're actually polling anything. Unless maybe `k8sGet` returns an error in which case it continues to poll? Is that it?

vikram-raj replied:
We cancel the interval when we get the status of the GitSource. we continue polling until we get the status of the GitSource.

christianvogt replied:
`validateGitRepo` creates the git source resource. It then starts an interval. After 3 seconds a call to `checkUrlValidationStatus` is made which immediately calls `k8sGet`. Since the resource was created 3 seconds ago, this will return successfully and then the poller interval is canceled. In the event of an error, then the interval makes sense. Are you encountering errors trying to retrieve the git source after the call to create the git source successfully returned? What kind of errors are you getting if that's the case?

debsmita1 replied:
@christianvogt  We arent encountering any errors after the create git source call. 
For your pervious comment, we stop polling the moment we get a validation status, be it reachable or unreachable. The polling starts again when the user modifies the url. I thought we cannot do this with the `watchk8sObject` service, so went ahead with polling.

christianvogt replied:
I get the reason for why you need to do something along these lines. I just don't see why the use of `setInterval` over `setTimeout`. I don't see where _interval_ polling is being used. Looks like it's a single delayed call to request the status and it's done only once per url update at which point the interval is restarted.

debsmita1 replied:
After creating Gitsource cr we are able to fetch the status in a single get request. But I figured out that on creating gitsourceanalysis, it takes more than a single get request to fetch the build type name. So considering this I thought its safe to keep polling at intervals till we get the validation status for a git url.

christianvogt replied:
You just said `We arent encountering any errors after the create git source call.`
So what happens when you make the first GET request? Looks to me like it will go into the `then` callback and clears the interval. Or does it simply not execute the callback and there are no errors report....?
(this is about url validation and not build type)

debsmita1 replied:
Now, whenever we make the first GET req, we are able to fetch the status and we clear the interval

vikram-raj committed:
fix(import): Add image stream dropdown and label to the component

debsmita1 committed:
feat(import-flow):auto detect build tool

vikram-raj committed:
fix(import): address review comment

divyanshiGupta commented on frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.scss:
```diff
@@ -0,0 +1,10 @@
+.odc-import-form {
+  &__success-icon {
+    color: #1e4f18;
```
Use pf variables for colors

debsmita1 replied:
Done : https://github.com/vikram-raj/console/pull/6
